src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -117,11 +124,61 @@ public Http2Connection(HttpConnectionPool pool, Stream stream)
             _maxConcurrentStreams = int.MaxValue;
             _pendingWindowUpdate = 0;
 
+            if (_pool.EnableMultipleHttp2Connections)
+            {
+                _freeStreamSlots = InitialSlotsLimit;
+            }
+
             if (NetEventSource.IsEnabled) TraceConnection(stream);
         }
 
         private object SyncObject => _httpStreams;
 
+        public bool AcquireStreamSlot()
+        {
+            while (true)
+            {
+                int currentSlots = _freeStreamSlots;
+                if (currentSlots == 0)
+                {
+                    long currentExpiration = _freeStreamSlotsExpiration;
+                    if (currentExpiration > Environment.TickCount64)
+                    {
+                        // No empty slots and it's too soon to request the latest value from the credit manager.
+                        return false;
+                    }
+
+                    _freeStreamSlotsExpiration = Environment.TickCount64 + SlotExpirationPeriod;
+
+                    int refreshedSlots = _concurrentStreams.Current;
+                    if (refreshedSlots == 0)
+                    {
+                        // Credit manager reported no available slots.
+                        return false;
+                    }
+
+                    if (Interlocked.CompareExchange(ref _freeStreamSlots, refreshedSlots, 0) != 0)
+                    {
+                        // Other thread already updated _freeStreamSlots, so let's try acquiring a free slot again.
```

Comment:
I doubt this will see enough contention to be measurable, but it would be slightly more efficient to set `currentSlots` to the return value of the CAS rather than re-read it at the top of the loop. Same with the CAS below.

Reply:
Good point. Done.